### ISSUE TITLE
PEP 585 usages of `list`, `tuple`, `type`, etc. in type aliases are not flagged when `requires-python` includes `< 3.9`.

### ISSUE THREAD

**Issue by jagerber48** [astral-sh/ruff#10290]
PEP 585 allows usages like `list[int]` to be used in type annotations and type aliases. In python 3.7 and 3.8 this syntax can be used in type annotations if paired with `from __future__ import annotations`. However, as far as I can tell, it is not possible to use this syntax in **type aliases** for these versions of python.

```
MyType = list[int]
```
gives 
```
TypeError: 'type' object is not subscriptable
```
on Python 3.8 for me.

Instead I find myself needing to do
```
from typing import List

MyType = List[int]
```

I am trying to extend support for a library that was previously `>= 3.9` to `>= 3.8`.
I expected ruff to give me linting warnings about the above patterns when I set `requires-python = ">= 3.8"` in `pyproject.toml` but instead I get no such warnings.

I tried using the `keep-runtime-typing` option but that didn't cause errors to be raised. I also tried ignoring `UP006` and `UP007`, I didn't expect that to work (since I would only expect that to cause ruff not to ASK me to switch from `List`  to `list`, but not force a `list -> List` back conversion.

Is there something I am missing?

I am using ruff version 0.3.1

**Comment by charliermarsh** [comment]
I don't think we currently flag those. We could, but AFAIK they're not covered by any rule right now. We only cover the inverse, e.g., you have a `List[int]` that could be a `list[int]`. It seems reasonable to support though.

**Comment by MichaReiser** [comment]
That makes sense to me. To me the question is how we want to scope this rule. Should it be a dedicated rule for possible typing errors because of version incompatibility or is it a generic rule that catches any possible incompatibilities based on your python version. I'm bringing this up because there are two similar problems:

* Ruff should warn about invalid syntax that raises a runtime `SyntaxError`
* Ruff should warn about syntax that is incompatible with your target Python version

**Comment by Skylion007** [comment]
We discussed adding pydowngrade rules that did the opposite of pyupgrade, if we do implement that might be a nice rule category to add `DOWN`.

**Comment by jagerber48** [comment]
Yes exactly. I was surprised that ruff let detectable runtime  errors like this through. 

And yes exactly, a DOWN category is what I was expecting. 

I guess it’s not obvious to me how the DOWN and UP rules would work together though. I guess the DOWN rules would only be relevant pursuant to the Python versions allowed, and likewise many other rules would need version checks. But maybe version checks are already in place at least for the upgrade direction.

**Comment by charliermarsh** [comment]
> But maybe version checks are already in place at least for the upgrade direction.

Yeah, that's correct -- the `UP` direction has version checks, we'd never change `List[int]` to `list[int]` on unsupported versions.

**Comment by zanieb** [comment]
@ods  You're using `from __future__ import annotations` there and I believe this issue is specifically about type aliases which you're not using there? We have a specific [`keep-runtime-typing`](https://docs.astral.sh/ruff/settings/#lint_pyupgrade_keep-runtime-typing) option for this use-case.

**Comment by ods** [comment]
@zanieb Thanks, didn't know about this option!

**Comment by FichteFoll** [comment]
Similarly, setting `target-version = "py38"` does not consider that even inside .pyi files it is not possible to use the union syntax or the stdlib types as generics when defining a `TypeAlias`, which is required afaik when targetting Python 3.8. For example:

```py
from typing import Dict, List, TypeAlias, Union

JSON: TypeAlias = Union[str, int, float, List[JSON], Dict[str, JSON]]  # raises UP006 and UP007 
JSON2: TypeAlias = "str | int | float | list[JSON2] | dict[str, JSON2]"  # raises PYI020
```

**Comment by MichaReiser** [comment]
@FichteFoll this, as far as I understand, is intentional. See https://github.com/astral-sh/ruff/issues/12256

**Comment by FichteFoll** [comment]
Thanks for the pointer. Will add a comment about my confusion there.

**Comment by asglover** [comment]
Just to clarify,  Is there no "DOWN" set of rules that I can use to lint a program to make sure that it won't commit syntax errors on the minimum allowable target-version? That's what I assumed the "target-version" setting did until a close reading. So it only guarantees not to suggest breaking upgrades? 

Is there are issue tracking "DOWN" rules I could +1 on?

Could a clarifying sentence be added to the documentation for "target-version"? 

I'd be happy to make the PR for the documentation change if you're okay with adding clarification. 

Thanks for making ruff!

**Comment by ntBre** [comment]
There is https://github.com/astral-sh/ruff/issues/2501 for a possible pydowngrade/DOWN rule set.

I'm not sure I'm following exactly on the `target-version` docs, so that might be good for a separate issue!

Thanks for the kind words!
